packages/service-provider-core/src/uri-generator.ts

```diff
@@ -142,7 +147,61 @@ function generateUri(options: CliOptions): string {
     validateConflicts(options);
   }
 
-  return `${Scheme.Mongo}${host || generateHost(options)}:${port || generatePort(options)}/${db || DEFAULT_DB}`;
+  return `${Scheme.Mongo}${host || generateHost(options)}:${port || generatePort(options)}${getDbAndQueryStringWithDirectConnectionIfRequired(dbAndQueryString || DEFAULT_DB)}`;
+}
+
+/**
+ * Parses a given mongodb:// connection string and adds the `directConnection=true` query parameter if required.
+ * See: https://github.com/mongodb/specifications/blob/master/source/connection-string/connection-string-spec.rst#reference-implementation
+ * @param uri mongodb:// connection string
+ */
+function addDirectConnectionQueryParameterToMongoUriIfRequired(uri: string): string {
+  const uriNoScheme = uri.substr(Scheme.Mongo.length);
+
+  // Split URI at first "/"
+  let splitIndex = uriNoScheme.indexOf('/');
+  if (splitIndex < 0) {
+    // maybe there's a question mark as separator
+    splitIndex = uriNoScheme.indexOf('?');
+  }
+
+  const userAndHostInfo = splitIndex < 0 ? uriNoScheme : uriNoScheme.substr(0, splitIndex);
+  const authDbAndOptions = splitIndex < 0 || splitIndex === uriNoScheme.length - 1 ? '' : uriNoScheme.substr(splitIndex);
+
+  // Check user and host informatino part to extract only hosts
+  const atIndex = userAndHostInfo.lastIndexOf('@');
+  const hostInfo = atIndex < 0 ? userAndHostInfo : userAndHostInfo.substr(atIndex + 1);
+  if (hostInfo.indexOf(',') > -1) {
+    // multiple hosts, i.e. a seed list is present -> return original uri
+    return uri;
+  }
+
+  if (authDbAndOptions) {
+    // Check if a replicaSet or directConnection parameter is already present
+    return `mongodb://${userAndHostInfo}` + getDbAndQueryStringWithDirectConnectionIfRequired(authDbAndOptions);
+  }
+  return `${uri.endsWith('/') ? uri : uri + '/'}?directConnection=true`;
```

Comment:
What happens if `directConnection=false` is explicitly specified?

Reply:
Then `authDbAndOptions` will be set and you'll return in line 181 which calls the `getDbAndQueryStringWithDirectConnectionIfRequired(...)` method which should take care of that. See this test case: https://github.com/mongodb-js/mongosh/pull/476/files/ee6dcba4f152178c047372ea42af0757fb07264f#diff-779e8029eb04be4763e9ba6da3c47a259c239920374b91ad80bb514db98987a2R100